### PR TITLE
Compatibility for Beat Saber v1.6.0 and updated to use BSIPA

### PR DESCRIPTION
**Pull request by Sarayalth**

### CODE REVIEW

File: BeatSaberHTTPStatus/BeatSaberHTTPStatusPlugin.csproj [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (opl-) [inline BeatSaberHTTPStatus/BeatSaberHTTPStatusPlugin.csproj]: Did your editor add a Byte Order Mark?

File: BeatSaberHTTPStatus/Plugin.cs [hunk 2]
```diff
@@ -50,11 +52,11 @@ public class Plugin : IPlugin {
 		public static readonly string GameVersion = "$BS_VERSION$"; // Populated by MSBuild
 
 		public string Name {
-			get {return "HTTP Status";}
+			get { return "HTTP Status"; }
 		}
 
 		public string Version {
-			get {return PluginVersion;}
+			get { return PluginVersion; }
```
Comment (opl-) [inline BeatSaberHTTPStatus/Plugin.cs]: Added unnecessary spaces.

File: BeatSaberHTTPStatus/Plugin.cs [hunk 3]
```diff
@@ -119,11 +126,12 @@ public class Plugin : IPlugin {
 				noteCutMapping.Clear();
 
 				statusManager.EmitStatusUpdate(ChangedProperties.AllButNoteCut, "menu");
-			} else if (newScene.name == "GameCore") {
+			}
+			else if (newScene.name == "GameCore") {
```
Comment (opl-) [inline BeatSaberHTTPStatus/Plugin.cs]: The brackets closing the previous and opening the next block should be on the same line.

File: BeatSaberHTTPStatus/Plugin.cs [hunk 4]
```diff
@@ -177,10 +198,10 @@ public class Plugin : IPlugin {
...
 				gameStatus.noteJumpSpeed = diff.noteJumpMovementSpeed;
 				gameStatus.songHash = level.levelID.Substring(0, Math.Min(32, level.levelID.Length));
-				gameStatus.songTimeOffset = (long) (level.songTimeOffset * 1000f / songSpeedMul);
-				gameStatus.length = (long) (level.beatmapLevelData.audioClip.length * 1000f / songSpeedMul);
-				gameStatus.start = GetCurrentTime() - (long) (audioTimeSyncController.songTime * 1000f / songSpeedMul);
-				if (practiceSettings != null) gameStatus.start -= (long) (practiceSettings.startSongTime * 1000f / songSpeedMul);
+				gameStatus.songTimeOffset = (long)(level.songTimeOffset * 1000f / songSpeedMul);
+				gameStatus.length = (long)(level.beatmapLevelData.audioClip.length * 1000f / songSpeedMul);
+				gameStatus.start = GetCurrentTime() - (long)(audioTimeSyncController.songTime * 1000f / songSpeedMul);
+				if (practiceSettings != null) gameStatus.start -= (long)(practiceSettings.startSongTime * 1000f / songSpeedMul);
```
Comment (opl-) [inline BeatSaberHTTPStatus/Plugin.cs]: A space should appear after the casts.

File: BeatSaberHTTPStatus/Plugin.cs [hunk 5]
```diff
@@ -252,7 +274,7 @@ public class Plugin : IPlugin {
 			}
 		}
 
-		private static T FindFirstOrDefault<T>() where T: UnityEngine.Object {
+		private static T FindFirstOrDefault<T>() where T : UnityEngine.Object {
```
Comment (opl-) [inline BeatSaberHTTPStatus/Plugin.cs]: Unnecessary space.

File: BeatSaberHTTPStatus/manifest.json [hunk 6]
```diff
@@ -0,0 +1,13 @@
+﻿{
+  "$schema": "https://raw.githubusercontent.com/beat-saber-modding-group/BSIPA-MetadataFileSchema/master/Schema.json",
+  "author": "opl-",
```
Comment (opl-) [inline BeatSaberHTTPStatus/manifest.json]: It's "opl", without the dash. As it turns out getting 3 character usernames is unfortunately nearly impossible on most websites nowadays. :/

File: protocol.md [hunk 7]
```diff
@@ -94,7 +94,7 @@ StatusObject = {
 		"staticLights": Boolean, // Static lights
 		"leftHanded": Boolean, // Left handed
 		"playerHeight": Number, // Player's height
-		"disableSFX": Boolean, // Disable sound effects
+		"sfxVolume": Number, // Disable sound effects
```
Comment (opl-) [inline protocol.md]: Please add the value range if possible. Something like this (meaning values range from 0 to 1, inclusively on both ends):

```suggestion
		"sfxVolume": Number, // Disable sound effects. [0..1]
```

File: BeatSaberHTTPStatus/BeatSaberHTTPStatusPlugin.csproj [hunk 8]
```diff
@@ -155,15 +162,13 @@
...
-        <![CDATA[
+      <Code Type="Fragment" Language="cs"><![CDATA[
           Log.LogMessage(string.Format("Replacing {0} in {1} to {2}", MatchExpression, InputFilename, ReplacementText));
           File.WriteAllText(
             OutputFilename,
             File.ReadAllText(InputFilename).Replace(MatchExpression, ReplacementText)
           );
-        ]]>
-      </Code>
+        ]]></Code>
```
Comment (opl-) [inline BeatSaberHTTPStatus/BeatSaberHTTPStatusPlugin.csproj]: Either keep the change and fix the indent of the code in this block, or get rid of the change entirely.

File: BeatSaberHTTPStatus/BeatSaberHTTPStatusPlugin.csproj [hunk 9]
```diff
@@ -155,15 +162,13 @@
...
     </Task>
   </UsingTask>
-</Project>
+</Project>
```
Comment (opl-) [inline BeatSaberHTTPStatus/BeatSaberHTTPStatusPlugin.csproj]: Files should end with an empty line.

File: BeatSaberHTTPStatus/BeatSaberHTTPStatusPlugin.csproj (hunk 1 above)
Comment (Sarayalth) [reply BeatSaberHTTPStatus/BeatSaberHTTPStatusPlugin.csproj]: yep, i think so

**Comment by Sarayalth** [comment]
Sorry about the styling, everything should be ok now

**Comment by opl-** [comment]
Almost! There are still two `else if`s and the BOM change. As far as I remember I had some issue with MSBuild and the BOM in the past, so I'd prefer that didn't change.

As far as testing goes, it seems I won't be able to handle that myself at least until tomorrow, so I'll try to look into how well BeatMods would handle release candidate versions. Might just let the community test it that way. :eyes:

**Comment by opl-** [comment]
Looks good! If I don't catch anything wrong during testing (when I can) it will get released.

Thank you!

**Comment by Sarayalth** [comment]
o7

**Comment by Sarayalth** [comment]
i just tested the changes to see if it works (i've been busy this days) and something might have changed with this game update, because as far as i could test something in OnNoteWasCut seems to fail with this error:
`ArgumentException: An item with the same key has already been added. Key: NoteCutInfo`

I think I fixed that ([code in dev branch on my fork](https://github.com/Sarayalth/beatsaber-http-status/tree/dev)), but combos and something else seems to be broken now :/

**Comment by Sarayalth** [comment]
Hey fixed that! 
I've been testing everything for a while at seems to work fine now

File: BeatSaberHTTPStatus/Plugin.cs [hunk 10]
```diff
@@ -353,11 +355,17 @@ public class Plugin : IPlugin {
...
 			}
+		}
+
+		public void OnNoteWasCut2(NoteController noteController, NoteCutInfo noteCutInfo) {
+			var gameStatus = statusManager.gameStatus;
+
+			var noteData = noteController.noteData;
 
-			List<CutScoreBuffer> list = (List<CutScoreBuffer>) afterCutScoreBuffersField.GetValue(scoreController);
+			List<CutScoreBuffer> list = (List<CutScoreBuffer>)afterCutScoreBuffersField.GetValue(scoreController);
```
Comment (opl-) [inline BeatSaberHTTPStatus/Plugin.cs]: Formatting.

File: BeatSaberHTTPStatus/Plugin.cs [hunk 11]
```diff
@@ -353,11 +355,17 @@ public class Plugin : IPlugin {
 					statusManager.EmitStatusUpdate(ChangedProperties.PerformanceAndNoteCut, "noteMissed");
 				}
 			}
+		}
+
+		public void OnNoteWasCut2(NoteController noteController, NoteCutInfo noteCutInfo) {
```
Comment (opl-) [inline BeatSaberHTTPStatus/Plugin.cs]: This is a very undescriptive name. How does it differ from `OnNoteWasCut`? Also, why is this one necessary and what order are they called in?

File: BeatSaberHTTPStatus/BeatSaberHTTPStatusPlugin.csproj [hunk 12]
```diff
@@ -40,13 +40,17 @@
...
-      <HintPath>$(GameDirPath)Beat Saber_Data\Managed\IllusionPlugin.dll</HintPath>
-      <Private>False</Private>
+    <Reference Include="IPA.Loader">
+      <HintPath>$(GameDirPath)\Beat Saber_Data\Managed\IPA.Loader.dll</HintPath>
+    </Reference>
+    <Reference Include="MainAssembly">
+      <HintPath>$(GameDirPath)\Beat Saber_Data\Managed\MainAssembly.dll</HintPath>
+    </Reference>
+    <Reference Include="Mono.Posix">
+      <HintPath>$(GameDirPath)\Beat Saber_Data\Managed\Mono.Posix.dll</HintPath>
```
Comment (opl-) [inline BeatSaberHTTPStatus/BeatSaberHTTPStatusPlugin.csproj]: What is this reference and why is it necessary?

File: BeatSaberHTTPStatus/BeatSaberHTTPStatusPlugin.csproj [hunk 13]
```diff
@@ -40,13 +40,17 @@
...
+      <HintPath>$(GameDirPath)\Beat Saber_Data\Managed\HMLib.dll</HintPath>
     </Reference>
-    <Reference Include="IllusionPlugin, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null">
-      <HintPath>$(GameDirPath)Beat Saber_Data\Managed\IllusionPlugin.dll</HintPath>
-      <Private>False</Private>
+    <Reference Include="IPA.Loader">
+      <HintPath>$(GameDirPath)\Beat Saber_Data\Managed\IPA.Loader.dll</HintPath>
+    </Reference>
+    <Reference Include="MainAssembly">
+      <HintPath>$(GameDirPath)\Beat Saber_Data\Managed\MainAssembly.dll</HintPath>
```
Comment (opl-) [inline BeatSaberHTTPStatus/BeatSaberHTTPStatusPlugin.csproj]: All of the above references need `<Private>False</Private>` (see below) so they won't be copied to the output directory.

File: BeatSaberHTTPStatus/manifest.json [hunk 14]
```diff
@@ -0,0 +1,13 @@
+﻿{
+  "$schema": "https://raw.githubusercontent.com/beat-saber-modding-group/BSIPA-MetadataFileSchema/master/Schema.json",
+  "author": "opl",
+  "description": "This plugin exposes information about the current game status, live over a WebSocket and over HTTP.",
+  "gameVersion": "1.6.0",
```
Comment (opl-) [inline BeatSaberHTTPStatus/manifest.json]: The versions here need to be populated by MSBuild.

**Comment by opl-** [comment]
Published as v1.9.0 and published on BeatMods and GitHub releases, as usual. Unfortunately couldn't do any comprehensive testing, but I trust that you tested it and that it works, especially considering the additional commits you made.

Thank you yet again for this PR and sorry for being so slow.

Oh, I should also mention that I did look into the whole release candidate thing: BSIPA supports semver version strings correctly and BeatMods seems to have no code that actually checks the format of the version string, so it should be possible. Still, I decided not to do it and decided to just release a patch if anything turns out to still be broken.

**Comment by Sarayalth** [comment]
Oh, sorry for not responding, i didn't see the notifications.

About the OnNoteWasCut2 thing, after some testing that did "work" and i thought it fixed everything, but apparently wasn't processing last cut info.

I fixed this after realising that with the change to BSIPA you don't need to subscribe to OnActiveSceneChange, it just works with a method with that name, but couldn't upload the commits as i've been without internet until just now.

I'll clean everything and upload in a bit, sorry about all this 🙇.

**Comment by Sarayalth** [comment]
Merged your changes with the fixes, tested looking at the score and rank, and it's working. Do I open another pr for this changes?

**Comment by opl-** [comment]
Yes, if you have any more fixes you should open a new PR explaining what issues they fix and how they do it.

**Comment by opl-** [comment]
Sorry, I read your explanation for `OnNoteWasCut2` again and I think I don't understand what you're trying to say. Does it work as is or is there currently some issue with it in v1.9.0?

**Comment by Sarayalth** [comment]
There's an issue with it in v1.9.0, i just sent a pr with a fix for it.

The thing is BSIPA don't need
`SceneManager.activeSceneChanged += this.OnActiveSceneChanged;`

And because i didn't know that at the time i started working on this i tested something and found that `OnNoteWasCut2` "worked", but some data was not processed correctly with that.

After realising what was the problem i fixed it and now it works as expected, processing correctly the score and rank of the last cut made.

**Pull request merged**